Content.Shared/Charges/Systems/SharedChargesSystem.cs

```diff
@@ -0,0 +1,62 @@
+using Content.Shared.Charges.Components;
+using Content.Shared.Examine;
+
+namespace Content.Shared.Charges.Systems;
+
+public abstract class SharedChargesSystem : EntitySystem
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<LimitedChargesComponent, ExaminedEvent>(OnExamine);
+    }
+
+    protected virtual void OnExamine(EntityUid uid, LimitedChargesComponent comp, ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        args.PushMarkup(Loc.GetString("limited-charges-charges-remaining", ("charges", comp.Charges)));
+        if (comp.Charges == comp.MaxCharges)
+        {
+            args.PushMarkup(Loc.GetString("limited-charges-max-charges"));
+            return;
+        }
+    }
+
+    /// <summary>
+    /// Tries to add a number of charges. If it over or underflows it will be clamped, wasting the extra charges.
+    /// </summary>
+    public void AddCharges(EntityUid uid, int change, LimitedChargesComponent? comp = null)
+    {
+        if (!Resolve(uid, ref comp, false))
+            return;
+
+        var old = comp.Charges;
+        comp.Charges = Math.Clamp(comp.Charges + change, 0, comp.MaxCharges);
+        if (comp.Charges != old)
+            Dirty(comp);
+    }
+
+    /// <summary>
+    /// Gets the limited charges component and returns true if there are no charges. Will return false if there is no limited charges component.
+    /// </summary>
+    public bool IsEmpty(EntityUid uid, LimitedChargesComponent? comp = null)
+    {
+        // can't be empty if there are no limited charges
```

Comment:
i'd say that it would be better if this is true, because otherwise it implies that there is a charge - if something has no charges, then it could be considered 'empty'

Reply:
also this otherwise allows emag system to bypass this check if the item that can do emag effects doesn't have a charge component - i'm not sure if you intended that or not

Reply:
its intended - see the admeme emag added